app/services/search/classified_listing.rb

```diff
@@ -0,0 +1,139 @@
+module Search
+  class ClassifiedListing
+    INDEX_NAME = "classified_listings_#{Rails.env}".freeze
+    INDEX_ALIAS = "classified_listings_#{Rails.env}_alias".freeze
+
+    class << self
+      def index(classified_listing__id, serialized_data)
+        SearchClient.index(
+          id: classified_listing__id,
+          index: INDEX_ALIAS,
+          body: serialized_data,
+        )
+      end
+
+      def find_document(classified_listing__id)
+        SearchClient.get(id: classified_listing__id, index: INDEX_ALIAS)
+      end
+
+      def create_index(index_name: INDEX_NAME)
+        SearchClient.indices.create(index: index_name, body: settings)
+      end
+
+      def delete_index(index_name: INDEX_NAME)
+        SearchClient.indices.delete(index: index_name)
+      end
+
+      def add_alias(index_name: INDEX_NAME, index_alias: INDEX_ALIAS)
+        SearchClient.indices.put_alias(index: index_name, name: index_alias)
+      end
+
+      def update_mappings(index_alias: INDEX_ALIAS)
+        SearchClient.indices.put_mapping(index: index_alias, body: mappings)
+      end
+
+      private
+
+      def settings
+        { settings: { index: index_settings } }
+      end
+
+      def index_settings
+        if Rails.env.production?
+          {
+            number_of_shards: 1,
+            number_of_replicas: 1
+          }
+        else
+          {
+            number_of_shards: 1,
+            number_of_replicas: 0
+          }
+        end
+      end
+
+      def mappings
+        # 1. "In Elasticsearch, there is no dedicated array datatype. Any field can
+        #    contain zero or more values by default, however, all values in the
+        #    array must be of the same datatype."
+        #
+        # https://www.elastic.co/guide/en/elasticsearch/reference/current/array.html
+
+        # 2. You don't have to specify type: "object" since it is the default.
+        #    Specifying type: "object" will break specs because when you later
+        #    call the mappings API on Elasticsearch, it will NOT return the
+        #    type: object key, value pair.
+        #
+        # https://www.elastic.co/guide/en/elasticsearch/reference/current/object.html#object
+
+        {
+          dynamic: "strict",
+          properties: {
+            id: {
+              type: "keyword"
+            },
+            author: { # Don't need to specify type: "object" - see comment 2 above
+              dynamic: "strict",
+              properties: {
+                username: {
+                  type: "keyword"
+                },
```

Comment:
why not have all of these one item hashes on line? to save the vertical space problem you mentioned I mean

Reply:
I find pretty-printed ES indexes a lot easier to read to be honest.

Reply:
@rhymes I hear that. I think there are a few things at the top of my mind about that:

- Even when I took the one item hashes to a single line, the method was still too long (38 lines).
- With the exception of object definitions, the Elasticsearch docs have it "prettified" so being consistent with the docs seems easier to follow in the future.
- Following that thread of consistency, if we start with this format of doing it this way and keep it consistent for all of Elasticsearch mappings, it's easier to understand in the future. It also makes it easier to refactor later if we choose to do that.
- It's certainly possible we update the mappings in the future to add more settings. So having it like this makes the code diff a little prettier when doing so. Similar to how:
```ruby
SOME_CONSTANT = [
  "some_string",
  "another_string",
].freeze
```
Having that comma on the last item in the Array prevents a code diff of adding a comma when another item is added later on - making future code diffs less noisy.

How do you feel about leaving it?